include/swift/AST/DiagnosticConsumer.h

```diff
@@ -111,6 +112,55 @@ class NullDiagnosticConsumer : public DiagnosticConsumer {
                         ArrayRef<DiagnosticArgument> FormatArgs,
                         const DiagnosticInfo &Info) override;
 };
+
+/// \brief DiagnosticConsumer that funnels diagnostics in certain files to
+/// particular sub-consumers.
+///
+/// Diagnostics that are not in one of the special files are emitted into every
+/// sub-consumer.
+class FileSpecificDiagnosticConsumer : public DiagnosticConsumer {
+public:
+  using ConsumerPair =
+      std::pair<std::string, std::unique_ptr<DiagnosticConsumer>>;
+
+private:
+  SmallVector<ConsumerPair, 4> SubConsumers;
+
+  /// A "map" sorted by the end locations of each file, so that a lookup by
+  /// position can be done using binary search.
+  ///
+  /// \see #consumerForLocation
+  SmallVector<std::pair<CharSourceRange, unsigned>, 4> LocationToConsumerMap;
+
+  /// Indicates which consumer to send Note diagnostics too.
+  ///
+  /// Notes are always considered attached to the error, warning, or remark
+  /// that was most recently emitted.
+  ///
+  /// If null, Note diagnostics are sent to every consumer.
+  DiagnosticConsumer *ConsumerForSubsequentNotes = nullptr;
+
+public:
+  /// Takes ownership of the DiagnosticConsumers specified in \p consumers.
+  ///
+  /// There must be no duplicate files, but an empty file name can be used to
```

Comment:
Because this sentence uses both "duplicate files" and "file name" it could potentially confuse a reader. How about "File paths must be unique, but an empty file path ..."?